miden-lib/asm/sat/internal/faucet.masm

```diff
@@ -0,0 +1,281 @@
+
+    # update the root of the TSMT containing the non-fungible assets
+    push.FAUCET_DATA_STORAGE_SLOT exec.account::set_item dropw
+    # => [ASSET]
+
+    # add the non-fungible asset to the input vault for asset preservation checks
+    exec.layout::get_input_vault_root_ptr movdn.4 exec.asset_vault::add_non_fungible_asset
+    # => [ASSET]
+end
+
+#! Burns a non-fungible asset associated with the non-fungible faucet the transaction is being
+#! executed against.
+#!
+#! Panics:
+#! - If the transaction is not being executed against a non-fungible faucet.
+#! - If the non-fungible asset being burned is not associated with the faucet the transaction is
+#!   being executed against.
+#! - If the non-fungible asset being burned does not exist or was not provided as input to the 
+#!   transaction via a note or the accounts vault.
+#!
+#! Stack: [ASSET]
+#! Outputs: [ASSET]
+#!
+#! - ASSET is the asset that was burned.
+proc.burn_non_fungible_asset
+    # assert that we are executing a transaction against the non-fungible faucet (access checks)
+    exec.account::get_id exec.account::is_non_fungible_faucet assert
+    # => [ASSET]
+
+    # duplicate asset
+    dupw
+    # => [ASSET, ASSET]
+
+    # fetch the root of the TSMT containing the non-fungible assets
+    push.FAUCET_DATA_STORAGE_SLOT exec.account::get_item
+    # => [TSMT_ROOT, ASSET, ASSET]
+
+    # prepare stack for removal of non-fungible asset from tracking TSMT
+    swapw padw
+    # => [ZERO, ASSET, TSMT_ROOT, ASSET]
+
+    # remove the non-fungible asset from the tracking TSMT
+    exec.smt::set
+    # => [OLD_VAL, TSMT_ROOT', ASSET]
+
+    # assert the `OLD_VAL` is not ZERO, indicating that the non-fungible asset exists.
+    # we only need to check ASSET[1] as this is always set to the faucet_id and can not be 0.
+    drop drop eq.0 not assert drop
+    # => [TSMT_ROOT', ASSET]
+
+    # update the root of the TSMT containing the non-fungible assets
+    push.FAUCET_DATA_STORAGE_SLOT exec.account::set_item dropw
+    # => [ASSET]
+
+    # remove the non-fungible asset from the input vault for asset preservation checks
+    exec.layout::get_input_vault_root_ptr movdn.4 exec.asset_vault::remove_non_fungible_asset
+    # => [ASSET]
+end
+
+# PUBLIC INTERFACE
+# ==================================================================================================
+
+#! Mint an asset from the faucet the transaction is being executed against.
+#!
+#! Panics:
+#! - If the transaction is not being executed against a faucet.
+#! - If the asset being minted is not associated with the faucet the transaction is being executed
+#!   against.
+#! - If the asset is not well formed.
+#! - For fungible faucets if the total issuance after minting is greater than the maximum amount 
+#!   allowed.
+#! - For non-fungible faucets if the non-fungible asset being minted already exists.
+#!
+#! Stack: [ASSET]
+#! Outputs: [ASSET]
+#!
+#! - ASSET is the asset that was minted.
+export.mint
```

Comment:
Actually, the more I think about it, the more it seems like the first approach is better.

Reply:
So I had initially implemented it as you described.  However, this results in a complication as it would prevent us from unifying the `mint` / `burn` procedures for fungible and non-fungible assets as they would take a different number of arguments as input (1 for the amount of fungible asset / 4 for the data hash of the non-fungible asset).  I concluded that having `ASSET` provided by the client was a good means of allowing us to unify the interface with the cost of delegating asset creation the client.  I like your proposal of introducing `asset.masm` in which we have methods for creating assets.  I will implement this.

Reply:
done.